File: dsa_leetcode_beginner/0253-meeting-rooms-ii/0253-meeting-rooms-ii.py
def minMeetingRooms(intervals):
    # Separate and sort start and end times
    start_times = sorted(interval.start for interval in intervals)
    end_times = sorted(interval.end for interval in intervals)

    start_pointer, end_pointer = 0, 0
    used_rooms = 0
    max_rooms = 0

    # Iterate over start times
    while start_pointer < len(intervals):
        if start_times[start_pointer] < end_times[end_pointer]:
            # A new room is needed
            used_rooms += 1
            start_pointer += 1
        else:
            # A room is freed up
            used_rooms -= 1
            end_pointer += 1
        
        # Update the maximum number of rooms needed
        max_rooms = max(max_rooms, used_rooms)

    return max_rooms


#Question: https://www.lintcode.com/problem/919/
#Blog: https://blog.unwiredlearning.com/meeting-rooms-ii

File: dsa_leetcode_beginner/0253-meeting-rooms-ii/0253-meeting-rooms-ii.cpp

#include <iostream>
#include <vector>
#include <algorithm>

int minMeetingRooms(std::vector<std::pair<int, int>> intervals) {
    std::vector<int> start_times, end_times;

    for (const auto& interval : intervals) {
        start_times.push_back(interval.first);
        end_times.push_back(interval.second);
    }

    std::sort(start_times.begin(), start_times.end());
    std::sort(end_times.begin(), end_times.end());

    int start_pointer = 0, end_pointer = 0, used_rooms = 0, max_rooms = 0;

    while (start_pointer < intervals.size()) {
        if (start_times[start_pointer] < end_times[end_pointer]) {
            used_rooms++;
            start_pointer++;
        } else {
            used_rooms--;
            end_pointer++;
        }
        max_rooms = std::max(max_rooms, used_rooms);
    }

    return max_rooms;
}

int main() {
    std::vector<std::pair<int, int>> intervals = {{0, 30}, {5, 10}, {15, 20}};
    std::cout << minMeetingRooms(intervals) << std::endl;
    return 0;
}
